internal/controllers/machineset/machineset_controller.go

```diff
@@ -969,16 +971,37 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
 
 func (r *Reconciler) reconcileUnhealthyMachines(ctx context.Context, cluster *clusterv1.Cluster, ms *clusterv1.MachineSet, filteredMachines []*clusterv1.Machine) (ctrl.Result, error) {
 	log := ctrl.LoggerFrom(ctx)
+
+	// Calculate how many in flight machines we should remediate.
+	// By default, we allow all machines to be in flight, unless otherwise.
+	maxInFlight := len(filteredMachines)
+	if strategy := ms.Spec.RemediationStrategy; strategy != nil {
+		var err error
+		maxInFlight, err = intstr.GetScaledValueFromIntOrPercent(strategy.MaxInFlight, int(ptr.Deref(ms.Spec.Replicas, 1)), true)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to calculate maxInFlight to remediate machines: %v", err)
+		}
+	}
+
 	// List all unhealthy machines.
 	machinesToRemediate := make([]*clusterv1.Machine, 0, len(filteredMachines))
 	for _, m := range filteredMachines {
+		if maxInFlight <= 0 {
```

Comment:
What about reducing maxInFlight by the number of machines already deleting, so we are not adding up deletions due to remediation and deletions due to rollout in case both happen at the same time.

Another thing we can do to make remediation and rollout working better togheter is to give higher deletion priority to machines yet to be remediated so e.g.

We are rolling out with maxUnavaivailabe 10, but there are 20 machines to be remediated and maxParallelism for remediation is 5.
- 5 machines will be remediated (remediation has priority on rollout)
- additional 5 machines can be deleted due to rolled out, however the deletion policy will pick up machines to be remediated first, giving the facto a boost to the remediation process

Reply:
I'd be cautious about introducing hard opinions here, e.g. blocking remediation on a scale down operation might arguably be a feature or a bug.

I do think having Machines needing remediation (MachineOwnerRemediatedCondition=false) to be included in out "isMachineHealthy" criteria for the DeletionPolicies makes sense. It will probably help to make most of the scenarios where controllers need to interoperate organically more efficient.